proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1204 @@
+               |   qtycls [ '(' '..' ')' | '(' qvar1 ',' ... ',' qvarn ')' ]
+               |   'module' modid
+
+
+      With ``-XLocalModules`` specified, expand this definition to be ::
+
+        export ::= qvar [ 'qualified' ]
+               |   qtycon [ '(' '..' ')' | '(' cname1 ',' ... ',' cnamen ')' ] [ 'qualified' ]
+               |   qtycls [ '(' '..' ')' | '(' qvar1 ',' ... ',' qvarn ')' ] [ 'qualified' ]
+               |   'module' [ 'qualified' ] modids [ exports ]
+
+      The nonterminal *exports* (from the `Haskell 2010 Report`_)
+      denotes an export specification: a list of ``export``.
+
+      We must now give these new export items semantics by saying how they
+      are interpreted.    
+
+   #. The interpretation of Haskell98-style export items (that is, without ``qualified``)
+      is unchanged::
+
+        interpretExpItem(haskell98item, ambient_env)
+          = { occ ↦ orig | modids.occ ∈ haskell98item
+                         , modids.occ ↦ orig ∈ ambient_env }
+
+      The only difference here from above is now we write ``modids`` for the qualifier.
+
+      Note that ``occ ↦ orig`` is returned, not the full qualified name. This is
+      unchanged from previously, but it was previously a requirement (in order to
+      be type-correct) to return an ``occ ↦ orig`` binding, while now the type
+      has been generalized to return a QEnv.
+      
+   #. If the export item is a ``qvar``, a ``qtycon``, or a ``qtycls`` (henceforth
+      called ``qual``) followed
+      by ``qualified``, the export
+      QEnv from the current module includes ``qual ↦ orig`` (where ``orig``
+      is the original name found by looking up ``qual`` in the ambient QEnv).
+
+      If ``qual`` has no qualifier, report a warning controlled by
+      ``-Wunqualified-qualified-exports``.
+
+      If there are bundled identifiers exported (in ``(...)``), these are
+      also exported with the same qualification as the root identifier,
+      even if the bundled identifiers are written with different qualifications.
+      (Example: the export item ``M.T(K1, M.K2, N.K3) qualified`` exports
+      ``M.T``, ``M.K1``, ``M.K2``, and ``M.K3``.)
+
+      This behavior is different than the behavior without the ``qualified``
+      keyword, which causes the unqualified name to be mapped to the original
+      name in the export environment (as is done prior to this proposal).
+
+      In code::
+
+        interpretExpItem(haskell98item@(modids0.occname(..)) 'qualified', ambient_env)
+          = { modids0.occ1 ↦ orig | modids1.occ1 ∈ haskell98item
+                                  , modids1.occ1 ↦ orig ∈ ambient_env }
+
+      The key part here is that the return is a qualified-name mapping. Note
+      that the qualification in the exported mapping is ``modids0``, not the
+      ``modids1`` that might be found in the list of bundled qualified names.
+      See the example a few paragraphs above.
```

Comment:
Please give a precise reference.  I don't know what example you are referring to

Reply:
I've clarified.